### PR TITLE
feat(postgresql): Add java.sql.Types.ARRAY support to PostgresqlDataT…

### LINKED ISSUE #646: Postgres Aarray Datatype not recognized
> **Migrated from SourceForge** — [Bugs #414](https://sourceforge.net/p/dbunit/bugs/414/)
> Originally reported by **maheshmurali** on 2018-01-18 10:39:37.299000
> Priority: `1`
> Assigned to *(unassigned)*
> SF status: pending

Not able to store Array Type data into postgres sql database. Getting the following error.

FAILED CONFIGURATION: @BeforeMethod setUp
org.dbunit.dataset.NoSuchColumnException: system.AREA -  (Non-uppercase input column: system_area) in ColumnNameToIndexes cache map. Note that the map's column names are NOT case sensitive.
       at org.dbunit.dataset.AbstractTableMetaData.getColumnIndex(AbstractTableMetaData.java:117)
       at org.dbunit.operation.AbstractOperation.getOperationMetaData(AbstractOperation.java:89)
       at org.dbunit.operation.AbstractBatchOperation.execute(AbstractBatchOperation.java:150)
       at com.siemens.sinecnms.integration.operation.policy.UpdatePolicyRestApi.setUp(UpdatePolicyRestApi.java:39)
       at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
       at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
       at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
       at java.lang.reflect.Method.invoke(Unknown Source)
       at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
       at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:523)
       at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:224)
       at org.testng.internal.Invoker.invokeMethod(Invoker.java:599)
       at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:869)
       at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1193)
       at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:126)
       at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
       at org.testng.TestRunner.privateRun(TestRunner.java:744)
       at org.testng.TestRunner.run(TestRunner.java:602)
       at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
       at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
       at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
       at org.testng.SuiteRunner.run(SuiteRunner.java:289)
       at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
       at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
       at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
       at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
       at org.testng.TestNG.runSuites(TestNG.java:1144)
       at org.testng.TestNG.run(TestNG.java:1115)
       at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:132)
       at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:230)
       at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:76)

---

**jeffjensen** — 2018-04-21 20:06:59.328000
Did you find the cause?

### PR DESCRIPTION
**Pull request by jeffjensen**
…ypeFactory

* Add ArrayType, recognized for every array column (e.g. integer[], text[]), read/written as PostgreSQL's own array literal text representation.
* On write, split the literal into top-level elements - honoring double-quoted, backslash-escaped elements and the unquoted NULL keyword - and bind via the standard Connection#createArrayOf(String, Object[]), so PostgreSQL itself parses and validates each element against the column's actual base type. Only single-dimension arrays are supported for writing.
* Document ArrayType in databases/postgresql.adoc and databases.adoc.

Refs: 646

Claude-Session: https://claude.ai/code/session_01XBjSJ2bgVtwqyxGv2nvwP9

## Summary by Sourcery

Add PostgreSQL array column support by introducing a dedicated ArrayType and wiring it into the PostgresqlDataTypeFactory for java.sql.Types.ARRAY.

New Features:
- Support PostgreSQL array columns (java.sql.Types.ARRAY) via a new ArrayType that round-trips values using PostgreSQL array literal text.
- Recognize array column types in PostgresqlDataTypeFactory and map them to ArrayType based on the reported SQL type name.

Enhancements:
- Document PostgreSQL array handling and ArrayType usage in the PostgreSQL database documentation.
- Record the new PostgreSQL array support feature in the project changelog.

Tests:
- Add unit tests for ArrayType covering type metadata, reading/writing literals, element parsing, and error cases.
- Add integration tests verifying PostgreSQL array columns (including nulls, empty arrays, and quoted elements) round-trip correctly through a real database.

## Summary by CodeRabbit

- **New Features**
  - Added support for PostgreSQL array columns, including null values, empty arrays, quoted and escaped elements, and null elements.
  - Supports reading arrays with any dimensionality; writing is supported for single-dimensional arrays.
  - Automatically recognizes PostgreSQL array types and preserves their SQL type information.

- **Documentation**
  - Updated PostgreSQL documentation with array support details, usage behavior, and known limitations.

### CODE REVIEW

File: src/test/java/org/dbunit/ext/postgresql/PostgresqlArrayIT.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
+package org.dbunit.ext.postgresql;
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/test/java/org/dbunit/ext/postgresql/PostgresqlArrayIT.java:1]: All 6 existing `*IT.java` files in `org.dbunit.ext.postgresql` (including `PostgresqlJsonIT.java`, this file's direct precedent) omit the LGPL header, starting straight at the package declaration. Adding it here would make this file the outlier rather than consistent with its siblings, so leaving it as-is to match the established convention for this specific file category.

**Comment by jeffjensen** [comment]
Addressed CodeRabbit's nitpick findings in fixup commit 59afa882 (not yet squashed):

* **Row order in `PostgresqlArrayIT`** — fixed. The round-tripped table is now read via `createQueryTable(testTable, "SELECT * FROM " + testTable + " ORDER BY ID")` instead of an unordered `SELECT *` through `createDataSet()`, so the row-indexed assertions no longer depend on incidental heap-scan order matching insertion order.
* **Backslash escapes outside quotes in `ArrayType`** — fixed. PostgreSQL's array-literal grammar allows escaping delimiter/brace characters directly in an unquoted element (e.g. `{a\,b}` is one element `a,b`, not two), which the parser previously mishandled since it only recognized `\` inside double quotes. Added the same escape handling to the unquoted branch, plus two regression tests (`testSetSqlValue_withBackslashEscapedDelimiterOutsideQuotes_treatsAsOneElement`, `testSetSqlValue_withBackslashEscapedBraceOutsideQuotes_treatsAsLiteralCharacter`).
* **`!Objects.isNull(_connection)` in `PostgresqlArrayIT`** — left as-is. 5 of the 6 existing `*IT.java` files in this package use this exact form, including the direct `PostgresqlJsonIT.java` precedent this file was modeled after; only one file uses `!= null`. Matching the dominant sibling convention over a purely-stylistic, low-value change.

Full unit suite (2091 tests) and the full `postgresql-16` IT suite are green against a live container after these changes.
